Разбиение на компоненты, добавление реактивности.

krivetko opened the pull request.

krivetko committed:
Разбиение на компоненты, добавление реактивности.

krivetko committed:
Небольшое исправление tsconfig

krivetko committed:
Еще один подход к конфигу

krivetko committed:
Stylelintrc fixed

krivetko committed:
Autofixes

krivetko committed:
Config changes

krivetko commented:
Не могу побороть тест с Stylelint, воспользовался рекомендацией Егора к одному из подобных вопросов - игнорировать его

krivetko committed:
Added ignoring directive to _id in base class
Block

mikebars1995 commented on package.json:
```diff
@@ -6,19 +6,30 @@
   "author": "",
   "license": "ISC",
   "devDependencies": {
     "@anikin/parcel-transformer-handlebars": "^1.0.9",
```
**Можно лучше**

Лучше задавать **фиксированные версии** зависимостям. Если удалить `package-lock.json` и попытаться повторно установить зависимости с помощью команды `npm install`, то **npm** установит последние версии зависимостей, что может привести к ошибкам.

Иногда его нужно удалить, так как возникают конфликты и проект не запускается

mikebars1995 commented on src/components/button/button.hbs:
```diff
@@ -1 +1 @@
-<button type="button" class="{{class}}">{{name}}</button>
\ No newline at end of file
+{{label}}
```
**Нужно исправить**

В проекте в некоторых файлах есть ошибка `EOF`. Для `git` важно наличие **пустой строки** в конце файла. Подробнее тут: https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log Можно в настройки IDE добавить `insert_final_newline = true`

mikebars1995 commented on static/css/auth.pcss:
```diff
@@ -32,19 +32,15 @@
     font-style: normal;
     font-weight: 500;
     font-size: 9px;
     color: #999999;
```
**Нужно исправить**


все **цвета** лучше вынести в `css`-переменные и использовать их, чтобы можно было мгновенно понимать, что это за цвет, и можно было бы подкорректировать его в одном месте  и не искать дублирования во всем коде

mikebars1995 commented on src/pages/register/register.hbs:
```diff
@@ -1,10 +1,8 @@
+<div class="register">
+    <h3 class="register__header">Регистрация</h3>
+    {{{register_form}}}
+    {{{button}}}
```
**Нужно исправить**


Кнопку сабмита нужно тоже убрать в форму (тег `form`), так как это элемент формы


У тега `form` можно установить событие `submit` при котором будет еще срабатывать клик на кнопке сабмита, которая должна быть помещена внутрь формы

mikebars1995 commented on src/pages/profile/profile.ts:
```diff
@@ -0,0 +1,73 @@
+      error_text: 'Введите корректный номер телефона',
+    }],
+};
+
```
**Нужно исправить**


* Элементы ошибок перерывают инпуты, после их отображения на инпут не получается кликнуть, скрин https://skr.sh/sIfgZVdLxUx

mikebars1995 commented on src/pages/profile/profile.ts:
```diff
@@ -0,0 +1,73 @@
+    },
+    {
+      label: 'Имя в чате',
+      name: 'nickname',
```
**Нужно исправить**


По заданию 1го спринта нужно давать инпутам **конкретные** атрибуты `name`, чтобы в 3м спринте не менять ничего при подключении к `Api`. Скрин из задания https://disk.yandex.ru/i/lSPqeGrAa20ayQ

mikebars1995 commented on static/index.html:
```diff
@@ -3,7 +3,7 @@
     <head>
```
**Нужно исправить**


Нужно установить корректный язык страницы `<html lang="ru">`

mikebars1995 commented on static/index.html:
```diff
@@ -3,7 +3,7 @@
-        <script type="module" src="../src/index.js"></script>
+        <script type="module" src="../src/index.ts"></script>
         <link rel="stylesheet" href="../src/index.pcss">
     </head>
```
**Нужно исправить**


Нужно ещё указать `viewport` для отзывчивости страницы
`<meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover">`
Так же есть еще ряд рекомендаций для `meta`, которые доступны [тут](https://github.com/thedaviddias/Front-End-Checklist#meta-tag)

mikebars1995 commented on src/utils/Block.ts:
```diff
@@ -0,0 +1,219 @@
+  //@ts-ignore
+  private _id: string;
+
+  protected props: any;
```
**Нужно исправить**



Тип **пропсов** внутри класса `Block` должен быть единым. Либо он должен приходить в дженерик (смотрите мой совет), либо нужно объявить его 1 раз снаружи и использовать везде

mikebars1995 commented on src/utils/Block.ts:
```diff
@@ -0,0 +1,219 @@
+import { v4 as makeUUID } from 'uuid';
+import { EventBus } from './EventBus';
+
+export default class Block {
```
**Можно лучше**

лучше всего передавать тип пропсов в виде дженерика в `Block`, чтобы можно было конкретный тип пропсов всегда иметь под каждый отдельный экземпляр класса `Block`. Будет очень универсально 

```js
abstract class Block<Props extends Record<string, any> = unknown> {

```

mikebars1995 commented on src/utils/HTTPTransport.ts:
```diff
@@ -0,0 +1,71 @@
+}
+
+export class HTTPTransport {
+  get = (url: string, options: Options) => this.request(url, { ...options, method: METHODS.GET }, options.timeout);
```
**Можно лучше**
Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище

```jsx

// создаем тип метода
type HTTPMethod = (url: string, options?: Options) => Promise<unknown>

export class HTTPTransport {

  // используем тип и удаляем дублирование в аргументах
  get: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.GET}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  put: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.PUT}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  post: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.POST}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  delete: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.DELETE}, options.timeout)
  )

```

mikebars1995 reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


Что сделано отлично:


* `Readme` хорошо оформлен
* Проект задеплоен и отлично работает
* Отлично, что не забыли про `.gitignore`
* Отлично, что ограничиваете версию `node`
* Хорошая структура папок и файлов
* Добавлен `ESLint` и `Stylelint`
* Отлично, что не забываете про `alt` в тегах `img`
* Отлично, что не забываете про тег `main` для главной части страницы

Что нужно исправить:

* В проекте в некоторых файлах есть ошибка `EOF`. Для `git` важно наличие **пустой строки** в конце файла. Подробнее тут: https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log Можно в настройки IDE добавить `insert_final_newline = true`
* все **цвета** лучше вынести в `css`-переменные и использовать их, чтобы можно было мгновенно понимать, что это за цвет, и можно было бы подкорректировать его в одном месте  и не искать дублирования во всем коде
* Кнопку сабмита нужно тоже убрать в форму (тег `form`), так как это элемент формы
* Элементы ошибок перерывают инпуты, после их отображения на инпут не получается кликнуть, скрин https://skr.sh/sIfgZVdLxUx
* По заданию 1го спринта нужно давать инпутам **конкретные** атрибуты `name`, чтобы в 3м спринте не менять ничего при подключении к `Api`. Скрин из задания https://disk.yandex.ru/i/lSPqeGrAa20ayQ 
* Нужно установить корректный язык страницы `<html lang="ru">`
* Нужно ещё указать `viewport` для отзывчивости страницы
* Тип **пропсов** внутри класса `Block` должен быть единым. Либо он должен приходить в дженерик (смотрите мой совет), либо нужно объявить его 1 раз снаружи и использовать везде





Что можно улучшить:

* Лучше задавать **фиксированные версии** зависимостям. Если удалить `package-lock.json` и попытаться повторно установить зависимости с помощью команды `npm install`, то **npm** установит последние версии зависимостей, что может привести к ошибкам.
* лучше всего передавать тип пропсов в виде дженерика в `Block`, чтобы можно было конкретный тип пропсов всегда иметь под каждый отдельный экземпляр класса `Block`. Будет очень универсально 
* Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище




Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

**Комментарии `Можно лучше` не обязательны к исправлению прямо сейчас, это рекомендации**

Удачного рефакторинга кода.

krivetko committed:
Внесены изменения по результатам ревью:
1. Добавлена пустая строка в конец файлов
2. Цвета в CSS оформлены переменными
3. Кнопки сабмита внесены в формы
4. Исправлено отображение информации об ошибках валидации - не идеально,
   но сейчас не мешает работе с интерфейсом. В планах переверстать форму
   для корректного отображения (или перейти на тултипы).
5. Проверены атрибуты инпутов, исправлена 1 ошибка (поле nickname
   изменено на display_name). Для аватара отдельного инпута нет - будет
   изменение по кнопке.
6. Установлен корректный язык страницы index.html
7. Указан viewport
8. Выполнен рефакторинг базового класса Block - добавлен дженерик для
   пропсов.

krivetko committed:
Дополнен viewport

krivetko committed:
Зафиксированы версии компонентов в package.json

mikebars1995 reviewed (commented):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
